server/src/main/java/org/elasticsearch/index/mapper/VersionEncoder.java

```diff
@@ -0,0 +1,256 @@
+         * e.g. "rc2" &lt; "rc11", "alpha523" &lt; "alpha1234"
+         */
+        HONOUR_NUMERALS
+        {
+            @Override
+            public void encode(String part, BytesRefBuilder result) {
+                prefixDigitGroupsWithLength(part, result);
+            }
+
+            @Override
+            public DocValueFormat docValueFormat() {
+                return DocValueFormat.VERSION_NUMERIC;
+            }
+        };
+
+        public abstract void encode(String part, BytesRefBuilder result);
+        public abstract DocValueFormat docValueFormat();
+
+        public static SortMode fromString(String mode) {
+            switch (mode) {
+                case "semver":
+                    return SEMVER;
+                case "numeric":
+                    return HONOUR_NUMERALS;
+                default:
+                    throw new IllegalArgumentException("Unknown version field mode: " + mode);
+            }
+        }
+
+    }
+
+    /**
+     * Encodes a version string given the given {@link SortMode}.
+     * First, the input version string is split into the following parts:
+     * <p>
+     * mainVersion(-preReleasePart)(+buildId)
+     *
+     */
+    public static BytesRef encodeVersion(String versionString, SortMode mode) {
+        //System.out.println("encoding: " + versionString);
+        if (legalVersionString(versionString) == false) {
+            throw new IllegalArgumentException("Illegal version string: " + versionString);
+        }
+        // extract "build" suffix starting with "+"
+        String buildSuffixPart = extractSuffix(versionString, BUILD_SEPARATOR);
+        if (buildSuffixPart != null) {
+           versionString = versionString.substring(0, versionString.length() - buildSuffixPart.length());
+        }
+
+        // extract "pre-release" suffix starting with "-"
+        String preReleaseId = extractSuffix(versionString, PRERELESE_SEPARATOR);
+        if (preReleaseId != null) {
+            versionString = versionString.substring(0, versionString.length() - preReleaseId.length());
+        }
+
+        // pad all digit groups in main part with numeric marker and length bytes
+        BytesRefBuilder encodedVersion = new BytesRefBuilder();
+        prefixDigitGroupsWithLength(versionString, encodedVersion);
+
+        // encode whether version has pre-release parts
+        if (preReleaseId != null) {
+            encodedVersion.append(PRERELESE_SEPARATOR_BYTE);  // versions with pre-release part sort before ones without
+            String[] preReleaseParts = preReleaseId.substring(1).split(DOT_SEPARATOR_REGEX);
+            boolean first = true;
+            for (String preReleasePart : preReleaseParts) {
+                if (first == false) {
+                    encodedVersion.append(DOT_SEPARATOR_BYTE);
+                }
+                boolean isNumeric = preReleasePart.chars().allMatch(x -> Character.isDigit(x));
```

Comment:
I wonder if we should treat the empty string as a numeric or as a string, does the spec say anything about this corner case?

Reply:
The specs say that "Identifiers MUST NOT be empty". Identifiers are the dot-separated parts, so I think we can check this in some form of validation already (e.g. check and error on consecutive dots like mentioned above)